include/exadg/postprocessor/time_control_statistics.cpp

```diff
@@ -0,0 +1,83 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2021 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+// ExaDG
+#include <exadg/postprocessor/time_control_statistics.h>
+#include <exadg/utilities/print_functions.h>
+
+namespace ExaDG
+{
+TimeControlDataStatistics::TimeControlDataStatistics()
+  : write_preliminary_results_every_nth_time_step(numbers::invalid_timestep)
+{
+}
+
+void
+TimeControlDataStatistics::print(dealii::ConditionalOStream & pcout, bool const unsteady) const
+{
+  time_control_data.print(pcout, unsteady);
+  if(Utilities::is_valid_timestep(write_preliminary_results_every_nth_time_step))
+    print_parameter(pcout,
+                    "Write preliminary results every nth time step",
+                    write_preliminary_results_every_nth_time_step);
+}
+
+void
+TimeControlStatistics::setup(TimeControlDataStatistics const & time_control_data_statistics_in)
+{
+  time_control_data_statistics = time_control_data_statistics_in;
+  time_control.setup(time_control_data_statistics.time_control_data);
+}
+
+bool
+TimeControlStatistics::needs_evaluation(double const           time,
+                                        types::time_step const time_step_number) const
+{
+  if(time_control.needs_evaluation(time, time_step_number))
+    return true;
+  if(write_preliminary_results(time_step_number))
+    return true;
+
+  return false;
+}
+
+unsigned int
+TimeControlStatistics::get_counter() const
+{
+  return time_control.get_counter();
+}
+
+bool
+TimeControlStatistics::write_preliminary_results(types::time_step const time_step_number) const
+{
+  if(Utilities::is_valid_timestep(
+       time_control_data_statistics.write_preliminary_results_every_nth_time_step))
+    if((time_step_number - 1) %
+           (time_control_data_statistics.write_preliminary_results_every_nth_time_step) ==
+         0 ||
+       time_control.reached_end_time())
```

Comment:
I think this is the wrong logic: consider the time interval t = [0, 50] for a simulation, with sampling performed during the interval t=[20,50] (a typical setting for turbulent flows). The present code also wants to write output for the time interval t=[0,20], which does not make sense. The logic should be like this

`if((time_control.needs_evaluation() and (time_control.get_counter() % write_preliminary_every_nth_evaluation == 0) or time_control.reached_end_time())`

Note the naming `nth_evaluation` instead of `nth_time_step`! `nth_evaluation` makes sure that we only write new output if the statistics evaluation has been invoked before in the same time step.

`time_control.needs_evaluation()` in the above if-statement makes sure that we only consider the interval t=[20,50].

The problem with this is that `needs_evaluation()` will again increment the counter inside `TimeControl`. This calls for incrementing the counter manually to avoid such side-effects.

Reply:
See also the comment here https://github.com/exadg/exadg/pull/262#discussion_r1001521153